More async calls

jtkech opened the pull request:
in progress ...

jtkech commented:
Not cool, `IRouter.GetVirtualPath()` is not async, so in `AutorouteRoute.GetVirtualPath()` there is this synchronous execution (`.Result` at the end) that uses e.g `contentManager.GetAsync()`.

    var displayRouteData = GetContentItemDisplayRoutes(context.HttpContext, contentItemId).Result;

Called to generate paths e.g through an urlHelper used in a view e.g to list all the blog posts.

In 3.0 when using the endpoint system another kind of urlHelper using a linkGenerator will be used, but with the same async problem. I will try to find a way to overcome this.

The problem is that we have a custom router for all autoroutes and we might have an autoroute for many content items, so we can't fully initialize this router once because we would have to get all content items to grab their `ContentItemMetadata.DisplayRouteValues`. So i'm thinking of some global route values shared by all content items unless their `contentItemId`, i will try it tomorrow.

jtkech commented:
So, i refactored the home and autoroute features to prevent some non async code executions.

- Because of a previous PR `SiteSettings.HomeRoute` is initialized to a non null value but this doesn't prevent another `ISiteService` implementation to return another type of `ISite` with a null `HomeRoute`. So here i added some null checks but there are other places without null checks, let me know.

    But i'm fine with the rule that any `ISiteService` would need to provide an `ISite` where `HomeRoute` is never null. If so, maybe i could remove my additional null checks as in other places, let me know.

- I thought about removing `HomeRoute` and just keep `Autoroute`, or making `Autoroute` dependent of `HomeRoute` and so on. But finally i kept `HomeRoute` and `Autoroute` and let them quite independent. `Autoroute` being a way to manage the home route setting, indeed it needs a feature that uses it but not forcibly `HomeRoute`. Then `HomeRoute` uses the home route setting but is not aware of `Autoroute`.

- I fixed some config ordering so that e.g the list of blog posts renders the right link for a blog post that is set as the homepage.

- I also fixed the fact that when we publish an item with another autoroute path, before the previous path was not removed from one of the 2 mapping tables.

jtkech commented:
Just tried to make the scripting engines async, but i reverted my changes because the `Jint.Engine` doesn't have an `ExecuteAsync()` method and we can't provide to it `GlobalMethod` that are async.

jtkech commented:
Most of the remaining non async calls are

- **In startup configurations**, that's ok.

- **In scripting methods**, maybe good that the `Jint.Engine` exposes an `ExecuteAsync()` so that we could provide to it async methods.

- **In `TaxonomyIndex.Map()`** using the new `_contentDefinitionManager.GetTypeDefinitionAsync()`. Maybe good that `YesSql` supports `MapAsync()` and `ReduceAsync()` for indexes.

    **Is it a problem if an index `Map()` uses itself the session?** Here through `GetTypeDefinitionAsync()` if not already cached and if we are using the database vesion of the definition store.

jtkech commented:
So, i made `FileContentDefinitionStore` thread safe by registering it as a tenant singleton and by using a `ReaderWriterLockSlim` to not lock readers if we are not writing the file.

But `ReaderWriterLockSlim` can't be used in an async context so here `FileContentDefinitionStore` now runs synchronously as before, knowing that definitions are cached.

Piedone commented on src/OrchardCore/OrchardCore.ContentManagement/FileContentDefinitionStore.cs:
```diff
@@ -1,60 +1,81 @@
+                _lock.ExitWriteLock();
             }
 
             return Task.CompletedTask;
```
I realize this wasn't changed now but still related: unless I'm misunderstanding here or this is also WIP there not much point in making such semi-async methods that actually run synchronously (it'll actually use slightly more resources). I'd also suggest not making methods async without a good reason and at least some implementations that actually run asynchronously since making everything async above it in the call stack is a huge impact.

~~In this case BTW WriteTo() could be used on ln 63.~~ No, because that'll actually throw a `NotImplementedException` :D. However, the file write can still be async (by using its `StreamWriter` directly for example. Similarly the file read in `LoadContentDefinitionAsync`.

jtkech replied:
@Piedone thanks for your feedback. Yes i agree, in this PR the main goal is to prevent most of `GetAwaiter().GetResult()` in a given request, maybe i did too much, need to be reviewed.

I made `IContentDefinitionStore` async because the service that uses it is async and it make sense with one implementation `DatabaseContentDefinitionStore` that do an async session query.

Here for `FileContentDefinitionStore` (now used by default), i 1st did it fully async with `.LoadAsync()` and `.WriteToAsync()`. Then i realized that it was not thread safe, so i used a `ReaderWriterLockSlim` that is not async so we can't use `await` that may do a thread switching.

For a better compromise, i will see if we could use more often the new `ValueTask` type that uses less allocations, but need to learn a little more on it.

Piedone replied:
Indeed, you're write about `ReaderWriterLockSlim`, didn't think about that. It seems [it gets really ugly once you mix it with async](https://stackoverflow.com/a/15883258/220230).

Since this file may be shared across instances in a multi-server environment (this is what happens with Azure App Services) I think a file-based locking mechanism similar to what we have in Orchard 1.x would work. While I'm uneasy by the idea actually I don't see any other simple and natural solution for this, and with `ReaderWriterLockSlim` there will still be concurrency issues.

jtkech replied:
Yes i agree. At some point, for the same reason i also used a file based lock to update a lucene index (in fact i used indirectly the lucene one), but @sebastienros said to me to not try to make working something that is not intended to work like this.

And he said the same for `FileContentDefinitionStore`, it is not intended to be used by multiple instance sharing the same file system, in this case we need to use `DatabaseContentDefinitionStore`.

jtkech replied:
That said, related to lucene, there is still this comment in the source code ;)

    // TODO: Lock over the filesystem in case two instances get a command to rebuild the index concurrently.

Skrypt replied:
hmm, I think there's a lot of refactoring to be made in the Lucene module. 😉

jtkech replied:
Anyway, i now use an async `SemaphoreSlim`, even a reader can async block another reader, because now `ContentDefinitionManager` also has a tenant level cache. So, for a given tenant the definition is loaded once, until the definition is updated again.

jtkech commented:
@sebastienros nothing urgent for me but here my little roadmap for this week (unless the last point).

- I think this one is ready, more async calls and some fixes (i could retrieve them if needed).

- A PR to update the usages of caching + tokens, not only to adapt them for distributed services, but because here i saw other things (updated here for 2 services) and i defined some rules e.g:

    - Always get a new token before consuming data.
    - Cancel the token after storing and flushing data (e.g if database).
    - Update the cache only when reading data from the store (not when storing).
    - Use the token to invalidate the cache.

    Here i needed to apply these rules e.g for `ContentDefinitionManager` but i also made all its usages async, so better to 1st merge what is already done before continuing this work on another PR.

- Then, merge dev in 3.0 and solve conflicts (i fixed the home route issue with preview9).

- That would be a good base for re-working on distributed services.

jtkech commented:
@sebastienros 

- Yes, making the content definition manager async hides other things but i think this PR is quite important. I made around 20 other calls async knowing that for some i kept the non async version for compatibility, this around recipes, razor pages, shapes, templates ... and fixed some issues.

- Yes, home route and autoroutes are not the same in 3.0 but here i introduced `AutorouteOptions` and `HomeRouteFeature` to make some code simpler that i planned to use in 3.0.

- **Do you want that i revert back the definition manager to be non async?** So that i could create its final version in the `caching` PR #4001, then this PR will be much more light, and both PRs will not depend on each other.

- **But i need a confirmation**, when using the database store, it was making sense for me to do session stuff async to prevent some db locks. Maybe even if we cache the definitions at the tenant level (currently it is at the scope level, only new built types / parts are cached at the tenant level). Note: and with distributed services where definitions could be invalidated ...

- Anyway the content definition manager needs to be fixed, a type / part can be built on a stale definition, there are some race conditions on concurrent mutations, if we keep it non async i can do that in the `caching` branch as i did for other things as `ISite`, `AdminMenu` and so on.

sebastienros commented:
I like it like that. Merge.

sebastienros commented:
once the tests pass obviously

sebastienros commented:
Also I am very surprised that the types were not cached correctly (as singleton), that should improve perf a lot.

sebastienros commented:
Also you can remove the non async Permission provider.

jtkech commented:
Okay thanks, will do but i want to redo some minimum tests.

Also i was just removing the `IScriptingEngine.EvaluateAsync()` (not committed yet). I did it async just for the `FilesScriptEngine` that is not frequently used (maybe only on setup).

So knowing that right now the `JavaScriptEngine` can't be async, do you think it's better that i continue to remove the related async calls?

jtkech commented:
> Also you can remove the non async Permission provider.

Only 2 uses the async version, so around 50 providers will return a `Task.FromResult`, that's ok?

> Also I am very surprised that the types were not cached correctly (as singleton), that should improve perf a lot.

Here i reverted all the changes around the `ContentDefinitionManager` so it is as before. This is in the `caching` branch that i committed the final version that caches the definitions at the tenant level.

Note: currently the definitions are cached at the scope level but each time we build a type / part (using the definitions) it is cached at the tenant level. The final version, committed in the `caching` branch #4001, also caches the base definitions (from which we build types / parts) at the tenant level.
